### PR TITLE
Add support for filtering extensions based on provided extension poin…

### PR DESCRIPTION
**Pull request by nitinja**
…t context

Adds extension points to each guest object. Also filters out extensions that are not matching to a provided extension point.

## Description

This PR is part of code, that handles host-side processing of `extensionPoint` associated with each extension.

- Added extensionPoint to each guest
- Filters out matching extensions
- 
## Related Issue
Extensions for CF Editor and CF Admin Header menu may be displayed in wrong app
https://jira.corp.adobe.com/browse/SITES-17283

## Motivation and Context
header menu buttons for CF Editor are rendered in the CF selector component (which contains header menu component as it is not moved to the level of the app)

## How Has This Been Tested?
Manually tested on local machine. 

## Screenshots (if appropriate):
N/A

## Types of changes

- [ X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ X] All new and existing tests passed.

### CODE REVIEW

**Comment by nitinja** [comment]
closing/reopening to re-run check for signing Adobe CLA.

File: packages/uix-core/src/types.ts [hunk 1]
```diff
@@ -100,6 +100,10 @@ export interface Extension {
    * An extension configuration
    */
   configuration?: Record<string, unknown>;
+  /**
+   * extension points for this extension
+   */
+  extensionPoints?: string[];
```
Comment (naydav) [inline packages/uix-core/src/types.ts:106]: Maybe is should be part of `configuration` atrribute...

File: packages/uix-host-react/package.json [hunk 2]
```diff
@@ -19,7 +19,7 @@
   "bugs": "https://github.com/adobe/uix-sdk/issues",
   "dependencies": {
     "@adobe/uix-core": "^0.8.4",
-    "@adobe/uix-host": "^0.8.4"
+    "@adobe/uix-host": "file:.yalc/@adobe/uix-host"
```
Comment (naydav) [inline packages/uix-host-react/package.json:22]: Need to remove `yalc` dependencies (and check package.lock)

Also, need to up versions of changes packages.

File: packages/uix-host-react/src/extensible-component-context.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import React, { createContext, useMemo } from "react";
+import type { PropsWithChildren } from "react";
+
+/**
+ * Context container with extensionPoint string.
+ *
+ * @internal
+ */
+export type ExtensibleComponentContextType = {
+  extensionPoint: string;
```
Comment (naydav) [inline packages/uix-host-react/src/extensible-component-context.tsx:25]: Required or optional?

File: packages/uix-host/src/host.ts [hunk 4]
```diff
@@ -395,6 +395,11 @@ export class Host extends Emitter<HostEvents> {
       const extensionConfiguration = isExtension(extension)
         ? extension.configuration
         : undefined;
+
+      const extensionPoints = isExtension(extension)
```
Comment (naydav) [inline packages/uix-host/src/host.ts:399]: It's not critical, but we can avoid double call of "isExtension"

File: packages/uix-host-react/src/hooks/useExtensions.ts [hunk 5]
```diff
@@ -131,6 +136,17 @@ export function useExtensions<
     const newExtensions = [];
     const guests = host.getLoadedGuests(requires);
     for (const guest of guests) {
+      //proceed only if this guest supports provided extension point
+      if (
+        currentComponentExtensionPoint &&
+        guest.extensionPoints &&
```
Comment (naydav) [inline packages/uix-host-react/src/hooks/useExtensions.ts:142]: Please add unit/integration tests or a Storybook to ensure that the logic (for all cases) is working as expected. 

For example, what will happen if currentComponentExtensionPoint is not defined (extensionPoint was not specified), but guest.extensionPoints contains some values? In this case the condition will no be satisfyed. Is it desire state?

**Review by naydav — CHANGES_REQUESTED** [review]

File: packages/uix-host-react/package.json (hunk 2 above)
Comment (vkublytskyi) [reply packages/uix-host-react/package.json:22]: Why do we use yalc within SDK? My assumption was we use yalc to link dev packages from different repositories

File: packages/uix-host-react/src/hooks/useExtensions.ts [hunk 6]
```diff
@@ -131,6 +136,17 @@ export function useExtensions<
...
+        !guest.extensionPoints.includes(
+          `aem/${currentComponentExtensionPoint}/1`
```
Comment (vkublytskyi) [inline packages/uix-host-react/src/hooks/useExtensions.ts:144]: Leaked abstraction: SDK must not be aware of AEM. Top-level product identifiers and versions cannot be hardcoded at this level.

**Review by vkublytskyi — CHANGES_REQUESTED** [review]

File: packages/uix-host-react/src/hooks/useExtensions.ts (hunk 6 above)
Comment (nitinja) [reply packages/uix-host-react/src/hooks/useExtensions.ts:144]: I forgot to remove hardcoding, will update code right away. I agree with SDK must not be aware of aem or any other user of library.

[the remark this answers is not in the document]
Comment (nitinja) [reply packages/uix-host-react/src/hooks/useExtensions.ts:144]: I forgot to remove hardcoding, will update code right away.

File: packages/uix-host-react/src/hooks/useExtensions.ts (hunk 5 above)
Comment (nitinja) [reply packages/uix-host-react/src/hooks/useExtensions.ts:144]: I will add testing support in project (currently there is none), so it will take some more time.

File: packages/uix-host-react/package.json (hunk 2 above)
Comment (nitinja) [reply packages/uix-host-react/package.json]: package `uix-host-react` uses `uix-host` as a dependency (which was, in my case, not updated automatically despite being in same monorepo) so I had to add this `uix-host` yalc dependency to `uix-host-react` package.

**Comment by slitviachenko** [comment]
@nitinja , Do these changes cover a scenario that @vkublytskyi [described](https://teams.microsoft.com/l/message/19:meeting_NTkyYmRlNjUtYjViOS00NmNjLWFmOGUtZDQxOTJiNDJlZDZl@thread.v2/1704469030610?context=%7B%22contextType%22%3A%22chat%22%7D) recently?
```
“admin of version 1” and “editor of version 2”
```

**Comment by nitinja** [comment]
@slitviachenko I believe this implementation supports the use case you mentioned above, but not at the same time.
if you mark a component as “admin of version 1”, extensions anywhere in it (unless overridden by another marked component) will be filtered exactly with "“admin of version 1”, no other extension will show up. 
We can NOT mark a component with more than one extensionPoint at this time.

**Comment by nitinja** [comment]
Implemented following review comments:
1. The wrapper component is now called `ExtensibleComponentBoundary`. This better explains the raison d'etre of the component that it means restriction on extensions.
2. Above component now accepts multiple "boundries", or extensionPoints - its much more flexible. 
3. Refactoring - extracted compare logic of extensions as a separate method.

File: packages/uix-host-react/src/extensible-component-context.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (kokoc) [inline packages/uix-host-react/src/extensible-component-context.tsx:1]: file naming is not consistent with other project

File: packages/uix-host-react/src/hooks/useExtensions.test.tsx [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      () => useExtensions<GuestApis, VirtualApi>(configFactory, []),
+      { wrapper }
+    );
+
+    expect(result.current.extensions.length).toBe(2);
+  });
+
+  test("returns filtered extensions when ExtensibleComponentBoundaryContext with extensionPoints value is provided with different version", () => {
+    const wrapper = ({ children }: { children: ReactNode }) => (
+      <ExtensibleComponentBoundaryContext.Provider
```
Comment (kokoc) [inline packages/uix-host-react/src/hooks/useExtensions.test.tsx:119]: Could you also cover `ExtensibleComponentBoundary`. Seems like it will be used by end users

**Review by kokoc — COMMENTED** [review]
The code looks good to me. Could you please add appropriate section to README where the main SDK docs live?

**Comment by nitinja** [comment]
Working on readme changes, will be committing soon

**Review by kokoc — APPROVED** [review]
